rust/pattern/constraint/type_/mod.rs

```diff
@@ -37,3 +37,15 @@ pub use relates::*;
 
 pub mod sub;
 pub use sub::*;
+
+use crate::common::token::Constraint::Abstract;
+use std::fmt;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct AbstractConstraint;
+
+impl fmt::Display for AbstractConstraint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", Abstract)
+    }
+}
```

Comment:
The `abstract` constraint doesn't have any payload. It's implemented as a unit struct that's wrapped in an `Option<>` in the `TypeVariable`: the idea is that the variable either has this constraint, or it doesn't.